[BEAM-6159] Migrate dataflow portable worker using shared library to process bundle

boyuanzz opened the pull request:
This is the PR in order to construct ExecutableStage and make dataflow protable batch worker use RemoteBundle to process bundle. There is an assumption in current implementation: there is only one input PCollection going to the ExecutableStage and all output PCollections will go to all following output receivers.

There will be a few following PRs:
1. Map output PCollection to following OutputReceivers.
2. Setup the gradle test task to track the new bundle processing  with ExecutableStage. 
3. Turn on the experiment path for streaming worker. 

R: @lukecwik 
cc: @huygaa11 @rohdesamuel

lukecwik commented on runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/LengthPrefixUnknownCoders.java:
```diff
@@ -74,6 +76,16 @@
           "kind:void");
 
   private static final String LENGTH_PREFIX_CODER_TYPE = "kind:length_prefix";
+  private static Set<Node> sdkToRunnerBoundaries = new HashSet<>();
```
You can't use statics because LengthPrefixUnknownCoders could be called from multiple threads.

lukecwik commented on runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/Nodes.java:
```diff
@@ -304,6 +305,29 @@ public String toString() {
+    @Override
+    public String toString() {
+      // The request may be very large.
+      return "ExecutableStagetNode";
```
ExecutableStagetNode -> ExecutableStageNode

lukecwik commented on runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/BatchDataflowWorker.java:
```diff
@@ -214,20 +214,23 @@ protected BatchDataflowWorker(
     // param/injection
     if (DataflowRunner.hasExperiment(options, "beam_fn_api")) {
-      Function<MutableNetwork<Node, Edge>, Node> sdkFusedStage =
+      final Boolean useSharedLibrary = DataflowRunner.hasExperiment(options, "use_shared_lib");
```
nit: you can drop final, and also Boolean -> bool

boyuanzz commented:
Run Dataflow PortabilityApi ValidatesRunner

boyuanzz commented:
Run Seed Job

boyuanzz commented:
Run Dataflow PortabilityApi ValidatesRunner

boyuanzz commented:
+R: @youngoli

huygaa11 commented on runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/BeamFnMapTaskExecutorFactory.java:
```diff
@@ -293,12 +333,45 @@ public Node typedApply(RemoteGrpcPortNode input) {
+        OutputReceiver[] outputReceivers = new OutputReceiver[Iterables.size(outputReceiverNodes)];
+        int index = 0;
+        Iterator<OutputReceiverNode> iterator = outputReceiverNodes.iterator();
+        while (iterator.hasNext()) {
```
I think there must be a method that turns iterator into an array.

boyuanzz replied:
I checked the doc talking about how to convert an iterator into a list. I found the under layer of implementation is copy the whole content into a immutable list. Then using this converting method will have the same time complexity and extra memory required, vs iterating on the iterator directly. Is there any concern about iterating it directly?

huygaa11 commented on runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java:
```diff
@@ -0,0 +1,436 @@
+
+  @Override
+  public Node apply(MutableNetwork<Node, Edge> input) {
+    RunnerApi.Components.Builder componentsBuilder = RunnerApi.Components.newBuilder();
```
Move it down to where it is used

boyuanzz replied:
fixed.

huygaa11 commented on runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java:
```diff
@@ -0,0 +1,436 @@
+
+    Map<Node, String> nodesToPCollections = new HashMap<>();
+    ImmutableMap.Builder<String, NameContext> ptransformIdToNameContexts = ImmutableMap.builder();
+    // A filed of ExecutableStage which includes the PCollection goes to worker side.
```
what is filed?

boyuanzz replied:
Sorry it's a typo. Changed to *field*.

huygaa11 commented on runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java:
```diff
@@ -0,0 +1,436 @@
+    Map<Node, String> nodesToPCollections = new HashMap<>();
+    ImmutableMap.Builder<String, NameContext> ptransformIdToNameContexts = ImmutableMap.builder();
+    // A filed of ExecutableStage which includes the PCollection goes to worker side.
+    Set<PCollectionNode> exexutableStageOutputs = new HashSet<>();
```
typo "exexutableStageOutputs"

boyuanzz replied:
fixed.

boyuanzz commented on .test-infra/jenkins/job_PostCommit_Java_ValidatesRunner_PortabilityApi_Dataflow.groovy:
```diff
@@ -28,7 +28,7 @@ PostcommitJobBuilder.postCommitJob('beam_PostCommit_Java_ValidatesRunner_Portabi
   description('Runs the ValidatesRunner suite on the Dataflow PortabilityApi runner.')
 
   // Set common parameters. Sets a 3 hour timeout.
```
This is the test only change, will be reverted.

boyuanzz replied:
Reverted

boyuanzz commented on runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateRegisterFnOperationFunction.java:
```diff
@@ -92,10 +92,12 @@
       BiFunction<String, String, Node> portSupplier,
-      Function<MutableNetwork<Node, Edge>, Node> registerFnOperationFunction) {
+      Function<MutableNetwork<Node, Edge>, Node> registerFnOperationFunction,
+      boolean use_shared_lib) {
```
In that way, the pipeline option needs to be passed into `CreateRegisterFnOperationFunction`. I feel like passing this boolean is easier for cleanup the old code path in the future.

boyuanzz commented:
ValidatesRunner tests report: https://scans.gradle.com/s/dscpw7tokgpbc/tests

boyuanzz commented on runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/BeamFnMapTaskExecutorFactory.java:
```diff
@@ -589,15 +662,14 @@ OperationNode createFlattenOperation(
    */
   static Function<Node, Node> createOutputReceiversTransform(
-      final String stageName, final CounterFactory counterFactory) {
+      final String stageName, final CounterFactory counterFactory, Network<Node, Edge> network) {
```
Deleted, thanks~

boyuanzz commented on runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/fn/control/ProcessRemoteBundleOperation.java:
```diff
@@ -0,0 +1,100 @@
+            public FnDataReceiver<?> create(String pCollectionId) {
+              return receivedElement -> {
+                for (OutputReceiver receiver : receivers) {
+                  LOG.info("[BOYUANZ LOG]: consume element {}", receivedElement);
```
fixed.

boyuanzz commented on runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/Nodes.java:
```diff
@@ -304,6 +305,29 @@ public String toString() {
+
+    public abstract ExecutableStage getExecutableStage();
+
+    public abstract Map<String, NameContext> getPTransformIdToPartialNameContextMap();
```
I leave it as blank for now, But this field may be used for sideinput, timer and states.

boyuanzz commented on runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateRegisterFnOperationFunction.java:
```diff
@@ -203,11 +209,22 @@ public CreateRegisterFnOperationFunction(
-        runnerNetwork.addEdge(registerFnNode, successor, HappensBeforeEdge.create());
+      if (this.use_shared_lib) {
+        for (Node predecessor : Sets.intersection(sdkSubnetworkNodes, runnerToSdkBoundaries)) {
+          predecessor = network.predecessors(predecessor).iterator().next();
```
done.

boyuanzz commented:
Run Java PreCommit

lukecwik commented on runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java:
```diff
@@ -0,0 +1,444 @@
+import org.apache.beam.vendor.grpc.v1_13_1.com.google.protobuf.ByteString;
+import org.apache.beam.vendor.grpc.v1_13_1.com.google.protobuf.InvalidProtocolBufferException;
+
+public class CreateExecutableStageNodeFunction
```
Add class comment.

lukecwik commented on runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/BatchDataflowWorker.java:
```diff
@@ -214,22 +215,34 @@ protected BatchDataflowWorker(
+      if (DataflowRunner.hasExperiment(options, "use_executable_stage_bundle_execution")) {
+        sdkFusedStage = new CreateExecutableStageNodeFunction(pipeline, IdGenerator::generate);
+        transformToRunnerNetwork =
+            new CreateRegisterFnOperationFunction(
```
Why do we need the CreateRegisterFnOperationFunction still, isn't registration handled by the RemoteBundle returned by the RemoteStageBundleFactory?

boyuanzz replied:
discussed offline. CreateRegisterFnOperationFunction creates the sdk fused graph which is also needed in constructing ExecutableStage.

lukecwik commented on runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateRegisterFnOperationFunction.java:
```diff
@@ -92,10 +92,12 @@
   public CreateRegisterFnOperationFunction(
```
I would have suspected that we would not have needed to change this.

boyuanzz commented on runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/BatchDataflowWorker.java:
```diff
@@ -214,22 +215,34 @@ protected BatchDataflowWorker(
-              this::createPortNode,
-              lengthPrefixUnknownCoders.andThen(sdkFusedStage));
-
+      if (DataflowRunner.hasExperiment(options, "use_executable_stage_bundle_execution")) {
```
once there is no more issue in batch mode, I'll have that follow up PR.

boyuanzz commented:
Run Java PreCommit

boyuanzz commented on runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/fn/control/ProcessRemoteBundleOperation.java:
```diff
@@ -0,0 +1,100 @@
+  public void start() throws Exception {
+    try (Closeable scope = context.enterStart()) {
+      super.start();
+      OutputReceiverFactory receiverFactory =
```
Fixed in this PR.

boyuanzz commented:
The squashed commit: https://github.com/apache/beam/pull/7015/commits/924bc07e9866c6dbd4b1cdcfb7d43838453defdf addressed most comments above. There are several issues left which will be in the following up PRs:
1. Map output PCollection to following OutputReceivers.
2. Setup the gradle test task to track the new bundle processing with ExecutableStage.
3. Turn on the experiment path for streaming worker.

Please take another look @lukecwik

boyuanzz commented:
Run Java PreCommit

lukecwik commented on runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java:
```diff
@@ -0,0 +1,449 @@
+
+/**
+ * Converts a {@link Network} representation of {@link MapTask} destined for the SDK harness into an
+ * {@link Node} containing {@link org.apache.beam.runners.core.construction.graph.ExecutableStage}.
```
nit: in a follow up PR:
```
 * Converts a {@link Network} representation of {@link MapTask} destined for the SDK harness into a
 * {@link Node} containing an {@link org.apache.beam.runners.core.construction.graph.ExecutableStage}.
```

lukecwik commented on runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/IntrinsicMapTaskExecutorFactoryTest.java:
```diff
@@ -174,8 +174,8 @@ public void testCreateMapTaskExecutor() throws Exception {
             null /* beamFnDataService */,
-            null, /* dataApiServiceDescriptor */
             null /* beamFnStateService */,
+            null,
```
nit: in a follow up PR add the comment stating what the `null` represents.

lukecwik commented on runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/SingleEnvironmentInstanceJobBundleFactory.java:
```diff
@@ -66,15 +68,21 @@ public static JobBundleFactory create(
-  private final IdGenerator idGenerator = IdGenerators.incrementingLongs();
+  private final IdGenerator idGenerator;
 
   private SingleEnvironmentInstanceJobBundleFactory(
```
nit: in a follow-up PR make all callers explicitly pass in their idgenerator instead of allowing idgenerator be null.

lukecwik commented on runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/environment/StaticRemoteEnvironment.java:
```diff
@@ -0,0 +1,66 @@
+  }
+
+  @Override
+  public void close() throws Exception {
```
nit: in a follow up PR, instead of using a lock object, make this method synchronized. Also prefer guard style statements:
```
if (isClosed) {
  return;
}
... do other stuff to perform close ...
```

lukecwik commented on runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/environment/StaticRemoteEnvironmentFactory.java:
```diff
@@ -0,0 +1,72 @@
+public class StaticRemoteEnvironmentFactory implements EnvironmentFactory {
+  public static StaticRemoteEnvironmentFactory forService(
+      InstructionRequestHandler instructionRequestHandler) {
+    StaticRemoteEnvironmentFactory factory = new StaticRemoteEnvironmentFactory();
```
nit: In a follow up PR, why not define and use a constructor that takes this parameter. This would allow you to make the member variable final.

lukecwik commented on runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/environment/StaticRemoteEnvironmentFactory.java:
```diff
@@ -0,0 +1,72 @@
+import org.apache.beam.sdk.fn.IdGenerator;
+
+/**
+ * An {@link EnvironmentFactory} that creates StaticRemoteEnvironment used by Dataflow runner
```
In a follow up PR.
This isn't specific to Dataflow, anyone should be able to use a an already existing InstructionRequestHandler if one chooses.
